src/SkyHope.LibraryManager.WebApi/Controllers/BookController.cs

```diff
@@ -1,12 +1,258 @@
+                return BadRequest("This book is not available for checkout.");
+            }
+
+            var userToAssign = await _userRepository.GetEntityAsync(user.UserId);
+            if (userToAssign is null)
+            {
+                return NotFound("Could not find user to assign to book.");
+            }
+
+            if (userToAssign.CheckedOutBooks.Count >= maxBookCount)
+            {
+                return BadRequest($"This user has {maxBookCount} books checked out, and may not check out more.");
+            }
+
+            if (userToAssign.LateFeeDue > 0)
+            {
+                return BadRequest($"This user has ${userToAssign.LateFeeDue} in late fees due and cannot check out another book until this is paid off.");
+            }
+
+            bookToUpdate.IsAvailable = false;
+            bookToUpdate.UserId = userToAssign.UserId;
+            bookToUpdate.User = userToAssign;
+            bookToUpdate.DueDate = DateTime.Now.AddDays(dueInDays);
+            userToAssign.CheckedOutBooks.Add(bookToUpdate);
+
+            return Ok();
+        }
+
+        [HttpPut]
+        public async Task<ActionResult> CheckInAsync(HttpBook book)
+        {
+            var bookToUpdate = await _bookRepository.GetEntityAsync(book.BookId);
+            if (bookToUpdate is null)
+            {
+                return NotFound();
+            }
+
+            bookToUpdate.IsAvailable = true;
+            bookToUpdate.User = null;
+            bookToUpdate.UserId = null;
+
+            if (bookToUpdate.DueDate < DateTime.UtcNow)
+            {
+                var daysLate = DateTime.UtcNow - bookToUpdate.DueDate.Value;
+                bookToUpdate.User.LateFeeDue = daysLate.Days * lateFeePerDay;
+            }
+
+            bookToUpdate.User.CheckedOutBooks.Remove(bookToUpdate);
+
+            return Ok();
+        }
+
+        [HttpPost]
+        public async Task<ActionResult> AddBookAsync(HttpModels.Book book)
+        {
+            var author = await _authorRepository.GetEntityAsync(book.AuthorId);
+            if (author is null)
+            {
+                return BadRequest("Author ID provided is not valid.");
+            }
+
+            var bookToAdd = new DataAccessBook
+            {
+                BookId = book.BookId,
+                Title = book.Title,
+                AuthorId = book.AuthorId,
+                DewyClass = book.DewyClass,
+                Isbn = book.Isbn,
+                Year = book.Year
+            };
+            if (!await _bookRepository.TryAddAsync(bookToAdd))
+            {
+                return StatusCode(StatusCodes.Status500InternalServerError, "An unexpected error occurred.");
+            }
+
+            return Created();
+        }
+
+        [HttpGet]
+        public async Task<ActionResult<HttpBook>> GetBooksByAuthorAsync(int authorId)
+        {
+            var results = new List<HttpBook>();
+            var allBooks = await _bookRepository.GetAllAsync();
```

Comment:
You're downloading all the books from the database and then filtering them, this is wasteful. Do the filtering as part of the database context - this will ensure the results are filtered before they're returned from the DB.
Make a method in the Repository called "GetBooksByAuthor" for example.